[Blueprint] Only show WordPress.org themes

BurakParsAydin opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes #57845  .

(For Bug Fixes) Bug introduced in PR # .

### Screenshots or screen recordings:
### Before
![Screenshot from 2025-05-12 22-49-02](https://github.com/user-attachments/assets/421bc85d-0303-4cba-a1d0-6a977b41a080)

### After
![Screenshot from 2025-05-13 14-32-32](https://github.com/user-attachments/assets/edca9002-78a5-4357-969e-1050dd8899fe)


<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->



<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install a few themes that are hosted on WordPress.org (e.g. "Twenty Twenty-Four"). 
2. Next, install a few themes that are not hosted on WordPress.org. You can use my testing themes for this purpose. Feel free to download them:
[MockPress.zip](https://github.com/user-attachments/files/20187304/MockPress.zip)
[CodeLite.zip](https://github.com/user-attachments/files/20187270/CodeLite.zip)
[PlainStarter.zip](https://github.com/user-attachments/files/20187271/PlainStarter.zip)
3. Go WooCommerce > Settings > Advanced . Check "Enable blueprint to import and export settings in bulk" option and save the changes. 
4. Now navigate to WooCommerce > Settings > Advanced > Blueprint (Beta) 
5. At this point, you should only see the themes that are hosted on WordPress.org

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

BurakParsAydin committed:
Filter out non-WordPress.org themes in Blueprint export

BurakParsAydin committed:
Fix the linting errors

BurakParsAydin committed:
Add the changelog file

chihsuan reviewed (commented):
Appreciate your contribution! @BurakParsAydin 

Let me clarify: We need to exclude themes that are not hosted on wordpress.org from appearing in the Blueprint export section (`/wp-admin/admin.php?page=wc-settings&tab=advanced&section=blueprint`).


<img width="448" alt="Screenshot 2025-05-12 at 11 25 44" src="https://github.com/user-attachments/assets/148c7de5-eb07-491e-93e3-8393aa70a718" />


That means we should to modify the `get_themes_for_export_group` method instead of `ExportInstallThemeSteps` class. The `ExportInstallThemeSteps` already has the logic to exclude non-WordPress.org themes by checking if `download_link` is available.

https://github.com/woocommerce/woocommerce/blob/33d3a7be645c9ef1e564e3b05ad042610b195523/plugins/woocommerce/src/Admin/Features/Blueprint/Init.php#L152-L175
Also, using the `ThemeURI` field to check if the theme is hosted on wordpress.org is not sufficient because some themes are hosted on wordpress.org but do not have a `ThemeURI` field. We should use the `download_link` field instead or other reliable source to check if the theme is hosted on wordpress.org.

For reference, here is the PR that addresses the plugins issue: https://github.com/woocommerce/woocommerce/pull/57807

Let me know if you have any questions. 🙂

BurakParsAydin committed:
Undo the last change

BurakParsAydin committed:
Filter installed themes to include only those hosted on WordPress.org

BurakParsAydin committed:
Fix the linting error

BurakParsAydin commented:
@chihsuan Thanks for your help! 🙂 I've updated the logic to ensure we only return themes that are hosted on WordPress.org. I’ve also adjusted my initial comment above to reflect the latest changes.

Regarding the PR: I needed to loop through each installed theme and call the wp_themes_api() function individually using the theme_information action. As far as I know, unlike the plugins API, the themes API doesn't support querying multiple slugs at once - it only accepts a single 'slug' per request. Passing multiple slugs results in a “themes API invalid” error.

Let me know if you'd like this result to be cached for performance, or if any further adjustments are needed!

Also, do you think I should add a transient and write a unit test for this, similar to what was done in the plugin PR you previously shared?

chihsuan commented:
Thanks for the update! @BurakParsAydin 

> As far as I know, unlike the plugins API, the themes API doesn't support querying multiple slugs at once - it only accepts a single 'slug' per request. Passing multiple slugs results in a “themes API invalid” error.

I think it should be possible to query multiple slugs at once. Here is the API URL: `https://api.wordpress.org/themes/info/1.2/?action=theme_information&request[slugs][]=twentytwentyone&request[slugs][]=twentytwentytwo`. As you can see, it returns the information for the two themes. Could you please check if this is the case?

> Let me know if you'd like this result to be cached for performance, or if any further adjustments are needed!

Yeah, I think we should cache the result for performance. You can test a site that has multiple themes installed and go to the blueprint page. We don't want to affect the page load time too much for the users.

chihsuan commented:
> The plugins_api() function returns information for all valid slugs and doesn’t throw an error if one of the plugins isn't hosted on WordPress.org—it simply skips the invalid ones and returns the rest. In contrast, themes_api() fails and returns an error if any of the provided slugs aren't hosted on WordPress.org, even if some of them are valid.


Thanks for the clarification. @BurakParsAydin Indeed, the theme seems to behave differently, making this a challenging issue. 🤔  I think we will need to investigate alternative approaches. Sending multiple requests to verify a theme's presence on WordPress.org is impractical, as it would result in a large number of requests, leading to performance issues.

BurakParsAydin commented:
@chihsuan I've been trying to find a solution for this, but I feel a bit stuck at this point.

@mikejolley If you have any insights on this issue, I'd really appreciate your input. You can check my latest comment above for more context, but to summarize: is there a way to use `themes_api()` to check whether a theme is hosted on WordPress.org in a single request, similar to how `plugins_api()` handles it?

The challenge I'm facing is that, unlike `plugins_api()`, the `themes_api()` function returns an error if any of the requested themes are not hosted on WordPress.org—even if the others are valid. This makes it tricky to handle multiple slugs in a single request, since a single invalid slug breaks the entire response.

chihsuan commented:
> The challenge I'm facing is that, unlike `plugins_api()`, the `themes_api()` function returns an error if any of the requested themes are not hosted on WordPress.org—even if the others are valid. This makes it challenging to handle multiple slugs in a single request, as a single invalid slug can break the entire response.

@BurakParsAydin I'm revisiting this PR and feeling uncertain about the optimal approach. I have an idea that we could try to filter out non-WordPress.org themes in the frontend code to avoid blocking the initial page load. However, I'm worried this might cause UI delays - we'd probably need to test this approach to assess its impact.

If you're keen to give it a try, that would be great. Otherwise, I think we could close this PR for now until we have a better solution. Thanks for your time and effort!

BurakParsAydin commented:
@chihsuan Thanks for the reply! I'm eager to work on this. Before I dive in, could you give me a hint about this?:

> we could try to filter out non-WordPress.org themes in the frontend code

chihsuan commented:
@BurakParsAydin I think we could try filtering the list of themes in JavaScript

The theme list is currently being fetched from `window.wcSettings?.admin?.blueprint_step_groups`. We could call the API asynchronously/or in parallel to check if the theme is available on `WordPress.org` by using useState + useEffect hooks.`

https://github.com/woocommerce/woocommerce/blob/3003db4b2c7d2c2a035bae055eeeecbd6bafa438/plugins/woocommerce/client/admin/client/blueprint/settings/slotfill.js#L40-L52

BurakParsAydin committed:
Merge branch 'trunk' into add/57845

BurakParsAydin committed:
Undo the last commit

BurakParsAydin committed:
Fix theme filtering logic in Blueprint export to exclude non-WP.org themes

chihsuan commented:
> CodeRabbit suggested four different points, but some of them seem inconsistent. Should I just apply the one about the transient?

Yes, we should ensure that this checking doesn't affect the page load time too much, so caching is a good idea.

BurakParsAydin committed:
Add transient for caching Blueprint export

chihsuan commented on plugins/woocommerce/src/Admin/Features/Blueprint/Init.php:
```diff
@@ -313,4 +325,52 @@ function ( $plugin ) use ( $api_response ) {
+			array(
+				'fields' => array(
+					'downloadlink' => true,
+				),
```
It seems that `$api_response` not only contains the download link but also other details about the theme. Could you please check if we could exclude other fields from the API response to reduce the response size?

<img width="1136" alt="Screenshot 2025-07-04 at 16 33 14" src="https://github.com/user-attachments/assets/c4c3a8b5-8a52-4a1f-a121-108c8d3121ce" />

chihsuan reviewed (commented):
Thanks @BurakParsAydin, nice work! 

I confirmed that non-wordpress.org themes are not showing up in the list. 👍 

Just a minor comment on the response fields, other than that, everything looks good.

BurakParsAydin committed:
Trim unused fields from theme API response to reduce response size

BurakParsAydin commented:
@chihsuan I trimmed the unused data. Some additional data is still included, but I don’t think there’s a way to disable it. The overall payload is small, though.
Thank you :slightly_smiling_face:

chihsuan committed:
Merge branch 'trunk' into add/57845

BurakParsAydin committed:
Check if theme key is a string before applying strtolower()

chihsuan reviewed (approved):
Thanks, LGTM
